### PR TITLE
Use fixed arch-specific device memory reserve

### PR DESCRIPTION
**Pull request by nluehr**
Replaces current 6% policy for device memory not claimed by BFC allocator
with fixed reservation size.

Default system_memory reserve sizes are:
*  500MB for Compute Capability <= 6.x (Pascal and earlier)
* 1050MB for Compute Capabilities of 7.x (Volta, Turing)
* 1536MB for Compute Capability >= 8.x (Ampere and later)

Also adds TF_DEVICE_MIN_SYS_MEMORY_IN_MB env var to directly override the
above defaults. This is needed in cases where external packages require
significant device memory allocations.

Using a fixed-sized system memory reserve can reduce the amount of device memory available to TF models on small-memory GPUs, such as Geforce cards. But it should eliminate the majority of out-of-memory cuda library initialization errors such as https://github.com/tensorflow/tensorflow/issues/44072, https://github.com/tensorflow/tensorflow/issues/43764, https://github.com/tensorflow/tensorflow/issues/42163, https://github.com/tensorflow/tensorflow/issues/41377, and https://github.com/tensorflow/tensorflow/issues/41196. 

Attn: @sanjoy

### CODE REVIEW

**Comment by kushanam** [comment]
@googlebot I consent.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 1]
```diff
@@ -906,24 +906,27 @@ Status VerifyVirtualDeviceSettings(
...
   //
   // If the available_memory is < 2GiB, we allocate 225MiB to system memory.
-  // Otherwise, allocate max(300MiB, kMinSystemMemoryFraction *
-  // available_memory) to system memory.
+  // Otherwise, depending on the capability version assign
+  //  500MiB (for cuda_compute_capability <= 6.x) or
+  // 1050MiB (for cuda_compute_capability <= 7.x) or
+  // 1536MiB (for cuda_compute_capability >= 8.x)
   //
   // In the future we could be more sophisticated by using a table of devices.
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc:918]: Is this last line still relevant?

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 2]
```diff
@@ -981,13 +984,36 @@ Status SingleVirtualDeviceMemoryLimit(const GPUOptions& gpu_options,
...
-    const int64 min_system_memory = MinSystemMemory(available_memory);
+    const int64 min_system_memory = MinSystemMemory(available_memory, cc_major);
     if (min_system_memory < allocated_memory) {
       allocated_memory -= min_system_memory;
     }
   } else {
     allocated_memory = total_memory * per_process_gpu_memory_fraction;
   }
+
+  // override the excluded memory when TF_DEVICE_MIN_SYS_MEMORY_IN_MB is set
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc:995]: Please use full sentences (start with uppercase, end with period).

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 3]
```diff
@@ -981,13 +984,36 @@ Status SingleVirtualDeviceMemoryLimit(const GPUOptions& gpu_options,
...
+  const char* force_device_reserved_bytes =
+      std::getenv("TF_DEVICE_MIN_SYS_MEMORY_IN_MB");
+  if (force_device_reserved_bytes != nullptr &&
+      strcmp(force_device_reserved_bytes, "") != 0) {
+    size_t allowable_reserved_memory;
+    sscanf(force_device_reserved_bytes, "%zu", &allowable_reserved_memory);
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc:1001]: Use ` safe_strto32`?

File: tensorflow/core/common_runtime/gpu/gpu_device.cc (hunk 2 above)
Comment (sanjoy) [reply tensorflow/core/common_runtime/gpu/gpu_device.cc:995]: Also later.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Thanks for the fix, this should be a good usability improvement!

Only minor comments, otherwise LGTM.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc (hunk 1 above)
Comment (nluehr) [reply tensorflow/core/common_runtime/gpu/gpu_device.cc:918]: I guess that's pretty much what we've implemented. So removed final comment line.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc (hunk 2 above)
Comment (nluehr) [reply tensorflow/core/common_runtime/gpu/gpu_device.cc:995]: Fixed.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc (hunk 3 above)
Comment (nluehr) [reply tensorflow/core/common_runtime/gpu/gpu_device.cc:1001]: Done. Also check for less than zero.

**Review by sanjoy — APPROVED** [review]

**Comment by sanjoy** [comment]
@nluehr The internal CI is failing and the failures look legitimate, can you PTAL?

**Comment by nluehr** [comment]
It should be working now, @sanjoy.

**Comment by gbaned** [comment]
Seems auto-merge is not happening but the changes are merged into master now, so we can close this. Thank you for the PR.

**Pull request closed without merging**